### DISCUSSION TITLE
Identifying the Root Cause of Duplicate Roles in JWT Token

### DISCUSSION THREAD

**Discussion by SSumithS** [keycloak/keycloak#35808] [Q&A]
Hi,
I’ve come across an issue where roles are being duplicated in the JWT token. 
When I decode the token, I notice that the "roles" section contains the same role multiple times.
 KeycloakRealmImport to import configurations into Keycloak. The JWT token is generated successfully, but some roles are listed more than once under the "roles" claim.

- Has anyone encountered this issue before? If yes, what was the root cause in your case?
- Could there be a configuration or setting that might be causing roles to duplicate in the token?
- What steps can I take to identify and fix the issue?

Sample:

{
    "aud": [
        "security-admin-console",
        "Test-management"
    ],
    "resource_access": {
        "realm-management": {
            "roles": [
                "view-Test",
                "Test-admin",
                "query-Test",
                "manage-Test"
            ]
        },
        "Test": {
            "roles": [
                "Test-user"
            ]
        },
        "account": {
            "roles": [
                "test-role"
            ]
        },
        "Data": {
            "roles": [
                "data-user"
            ]
        }
    },
    "scope": "openid profile email",
    "sid": "",
    "email_verified": false,
    "org": "Test",
    "roles": [
        "view-Test",
        "Test-admin",
        "query-Test",
        "manage-Test",
        "Test-user",
        "test-role",
        "data-user",
        "view-Test",
        "Test-admin",
        "query-Test",
        "manage-Test",
        "Test-user",
        "test-role",
        "data-user",
        "view-Test",
        "Test-admin",
        "query-Test",
        "manage-Test",
        "Test-user",
        "test-role",
        "data-user"
    ],
    "groups": [
        "Administrators",
        "Users"
    ],
    "preferred_username": "TestUser"
}

**Comment by danielFesenmeyer** [comment]
Maybe the duplicated roles are assigned to the user in different ways (directly, via group, via other role, ...)?

But even in that case, without information on the type of assignment, the duplication does not provide helpful information.

**Reply by SSumithS** [reply]
Keycloak is deployed along with KeycloakRealmImport, and the Keycloak database is also set up. All configurations for realms, clients, users, and roles are specified using KeycloakRealmImport. This is the procedure i followed. sharing the sample yaml format( not exact one)


apiVersion: keycloak.org/v1alpha1
kind: KeycloakRealmImport
metadata:
  name: test-realm-import
  namespace: keycloak
spec:
  realm: Test
  enabled: true
  users:
    - username: TestUser
      emailVerified: false
      email: testuser@example.com
      firstName: Test
      lastName: User
      groups:
        - Administrators
        - Users
      clientRoles:
        security-admin-console:
          - view-Test
          - Test-admin
          - query-Test
          - manage-Test
        Test:
          - Test-user
        account:
          - test-role
        Data:
          - data-user
      realmRoles:
        - view-Test
        - Test-admin
        - query-Test
        - manage-Test
        - Test-user
        - test-role
        - data-user
      attributes:
        aud:
          - security-admin-console
          - Test-management
        scope:
          - openid
          - profile
          - email
      preferredUsername: TestUser
  roles:
    - name: view-Test
      description: View access to Test resources
    - name: Test-admin
      description: Admin access to Test resources
    - name: query-Test
      description: Query access to Test resources
    - name: manage-Test
      description: Manage access to Test resources
    - name: Test-user
      description: User access to Test resources
    - name: test-role
      description: Basic user role
    - name: data-user
      description: Access to Data resources
  groups:
    - name: Administrators
      realmRoles:
        - view-Test
        - Test-admin
        - query-Test
        - manage-Test
    - name: Users
      realmRoles:
        - Test-user
        - data-user
